### PR TITLE
NLLB-200 / M2M-100 (CPU + CUDA + Metal) + PEFT LoRA adapters [stacked on #321]

### PR DESCRIPTION
**Pull request by rsafier**
## NLLB-200 / M2M-100 (encoder-decoder) support + PEFT LoRA adapters

**Stacked on #321** (`feat/lora-on-hss-tier`). The PR base is `feat/lora-on-hss-tier`, so the diff reviews as **only the NLLB delta** (4 commits / 67 files), independent of #321.
The **incremental scope of this PR** is:

1. **NLLB-200 / M2M-100 runtime** (folds in MonumentalSystems#51, 16 commits incl.
   the latest metal beam-batching): a self-contained seq2seq encoder-decoder path.
   Atlas's production engine is decoder-only + GPU-only, so NLLB gets:
   - `crates/spark-nllb`: CPU fp32 reference (greedy + beam), bit-exact to HF.
   - CUDA (GB10) + Metal path in `spark-model` (`examples/nllb_cuda_*`,
     `examples/nllb_metal_*`, `kernels/{gb10,metal}/nllb-200-3.3b` +
     `common/nllb_encoder.{cu,metal}`): full encoder+decoder, cross-attn, KV cache,
     beam search, bf16 tensor-core decode, M=1 GEMV decode, request/beam batching.
   - `spark-model` marker: recognizes `model_type = "m2m_100" | "nllb"`.

2. **PEFT LoRA adapter support for NLLB** (new): runtime low-rank delta
   `y = x·Wᵀ + b + scale·(x·Aᵀ)·Bᵀ` (never merged) on **every** projection —
   encoder+decoder self-attn `q/k/v/out_proj`, decoder cross-attn `q/out_proj` +
   cached cross `k/v_proj`, and FFN `fc1/fc2`. Standard HF PEFT format
   (`adapter_config.json` + `adapter_model.safetensors`); `--lora <dir>` CLI flag +
   `NllbModel::load_dir_with_lora`. Covers both greedy and beam decode. See
   `crates/spark-nllb/src/lora.rs`.

### Testing (run before opening this PR)

- **7 spark-nllb unit tests** (delta math, zero-B no-op, PEFT key stripping,
  load round-trip, config null-tolerance) — green.
- **Gated e2e vs real `facebook/nllb-200-3.3B`** (`NLLB_MODEL_DIR`): base greedy
  **and** beam-5 still **exact-token-match to HF**, a zero-B adapter is a
  byte-exact no-op, a live adapter changes the output — 4/4 green.
- **Real production adapter**: downloaded mobtranslate's **Kuku-Yalanji v21.2
  "Atlas dynamic LoRA bundle"** (base `nllb-200-distilled-1.3B` + PEFT adapter,
  r=64/α=128, all 6 target module types) and translated `eng_Latn → gvn_Latn`.
  All **384 modules** loaded; the adapter demonstrably corrects the output:

  | English | base (no LoRA) | + Kuku adapter |
  |---|---|---|
  | The children are eating fish. | `Kanbalda kaykay nukan-nukajiku.` | `Karrkay-karrkay kuyu nuka-l-nuka.` |
  | I am going to the river tomorrow. | `Ngayu wawu-wayjujiku rivermunbu dunganka.` | `Ngayu wawubaja-nga dungay wunkun.` |

  (Needed one config fix: real distilled-NLLB checkpoints ship `"max_length": null`,
  which the parser now tolerates.)

- `cargo check -p spark-model --features cuda` green; `spark-nllb` fmt + clippy
  clean; all touched files ≤500 LoC.

### Follow-up
LoRA is wired into the CPU runtime; the GPU (CUDA/Metal) NLLB path does not yet
apply adapters — this crate's CPU delta is the reference for that.

### CODE REVIEW

**Comment by rsafier** [comment]
### Major addition: NLLB now serves through `spark-server` like any flagship model

The PR grew well beyond the CPU runtime + CPU LoRA. NLLB-200 / M2M-100 is now a
first-class served model — **`spark serve --model <nllb-dir>`** translates through
the **same** scheduler, OpenAI API, sampling and LoRA stack as Qwen/Holo. No
parallel server, no hardcoded paths (weights load from the standard `--model`
store).

**How** (`crates/spark-model/src/model/nllb/`): the validated bf16 CUDA
encoder-decoder runtime was promoted from the examples into a proper `Model` impl
(`NllbGpuModel`). It runs the encoder inside `prefill` (seeding the decoder with
`[decoder_start, forced_bos]` so the forced target-lang token stays out of the
sampled stream), `forward_one` per decode step, owns all its KV (empty
`block_table`, large `num_free_blocks`), and returns bf16 logits — the scheduler's
default sampling path, no fp32 overrides. `factory::build_model` branches to it
for `m2m_100`/`nllb`, bypassing the decoder-only loader; `preflight` skips the
encoder-decoder layout; a MODEL.toml size-wildcard lets the distilled-1.3B share
the dim-agnostic kernels.

**Features, all validated on real checkpoints:**
- **Base translate** — token-identical to the CPU reference.
- **LoRA** — reuses `apply_lora_delta` + `LoraPair` (no new kernels); the real
  mobtranslate **Kuku-Yalanji v21.2** adapter (384 modules, r=64) loads and shifts
  output correctly, token-identical to the CPU reference.
- **Batched decode (C=8)** — contiguous `[n,vocab]` logit rows, per-slot KV.
- **Per-request LoRA** — `adapter` request field → apply/skip per sequence.
- **Per-request language** — new `src_lang`/`tgt_lang` request fields (threaded
  like `adapter_slot` through `SequenceState`); one endpoint serves many languages.

**The C=8 demo** — one endpoint, 8 concurrent requests mixing 4 target languages
× (base + LoRA):
```
deu base → Die Kinder essen Fisch.          deu LoRA → Die Kinder fressen Fisch.
fra base → Les enfants mangent du poisson.  spa LoRA → Los niños comiendo peces.
gvn base → Kanbalda kaykay nukan-nukajiku.  gvn LoRA → Karrkay-karrkay kuyu nuka-l-nuka.
```

**No regression:** Holo-3.1-35B-A3B (MoE flagship) and Holo-3.1-0.8B (dense) both
serve correctly on the final binary — the shared-code changes (2 additive
`SequenceState` fields, an NLLB-only factory branch, NLLB-only preflight skip) are
inert for decoder-only models.

**Cleanup:** the redundant CUDA/Metal example harnesses (`nllb_cuda_*`,
`nllb_metal_*`) are removed now that the serve path supersedes them. The
`spark-nllb` **CPU fallback** crate + CLI + its bit-faithful **reference tests** are
retained as the correctness oracle and the non-GPU path.

CI-parity verified locally: `cargo fmt --check`, ≤500-LoC cap, `clippy -D warnings`
(spark-model + spark-server), rustdoc, non-cuda compile, and spark-model/spark-nllb
tests all green.

**Comment by rsafier** [comment]
### Added: batched beam search (`num_beams=5`, NLLB default decoder)

Beam search — NLLB's default, best-quality decoder — now serves through the same
OpenAI API. It's a run-to-completion path (best hypothesis only known at the end),
so it does NOT fit the scheduler's token-by-token loop; instead a new `Model`
seam runs it inside prefill:

- `Model::supports_beam()` (default false; only `NllbGpuModel` = true) +
  `generate_beam_batch(&[BeamReq])`. The prefill path (chunked + legacy) detects
  `supports_beam() && num_beams>1`, runs the whole beam search, and emits the
  winning hypothesis as a **finished** sequence — no decode-loop edits, greedy/
  Qwen/Holo byte-identical.
- **Batched compute**: all B beams decode as ONE M=B bf16 tensor-core GEMM per
  projection (reusing the encoder's multi-row `linear`, which also applies
  per-request LoRA at M=B), batched attention with per-beam key lengths, shared
  cross-KV (stride 0, no replication), on-device scatter/gather for KV write +
  beam reorder. Kernels already in-tree (`nllb_encoder.cu`). This is the
  milestone-7 "2.6×" design.
- **API**: per-request `num_beams` / `length_penalty` / `early_stopping` (mirror
  the `src_lang`/`tgt_lang` plumbing); `stream=true` + `num_beams>1` → 400.

**Validated:** served beam=5 is token-exact vs the CPU reference, genuinely
differs from greedy on longer inputs (real beam search, not a fallthrough), and
runs in **~1 s/request** (vs the CPU beam >2 min). Composes with per-request LoRA
and per-request language; Holo-3.1-0.8B/35B greedy unregressed.

**Next (perf, optional):** cross-request C×B co-dispatch (one `generate_beam_batch`
over all pending beam requests → a single `[C·B,d]` forward) + on-device top-k to
remove the per-step host D2H — for high-concurrency (C=64/128) beam throughput.
Today concurrent beam requests batch internally at B but serialize per-request on
the scheduler thread.
